File: java/src/test/java/com/example/FirebaseAdminTest.java
package com.example;

import com.google.auth.oauth2.GoogleCredentials;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.ListUsersPage;
import com.google.firebase.cloud.FirestoreClient;
import com.google.cloud.firestore.DocumentReference;
import com.google.cloud.firestore.DocumentSnapshot;
import com.google.cloud.firestore.Firestore;
import org.junit.jupiter.api.*;

import java.io.IOException;

import static org.junit.jupiter.api.Assertions.*;

@TestMethodOrder(MethodOrderer.MethodName.class)
public class FirebaseAdminTest {

    private static FirebaseApp app;

    @BeforeAll
    public static void setup() {
        // Initialize the app via ADC
        if (FirebaseApp.getApps().isEmpty()) {
            try {
                FirebaseOptions options = FirebaseOptions.builder()
                        .setCredentials(GoogleCredentials.getApplicationDefault())
                        .setServiceAccountId("firebase-adminsdk-fbsvc@admin-sdk-wif.iam.gserviceaccount.com")
                        .setProjectId("admin-sdk-wif")
                        .build();
                app = FirebaseApp.initializeApp(options);
            } catch (IOException e) {
                throw new RuntimeException("Failed to initialize FirebaseApp with ADC credentials", e);
            }
        } else {
            app = FirebaseApp.getInstance();
        }
    }

    @Test
    public void test1InitializeApp() {
        assertNotNull(app);
        System.out.println("Firebase Admin SDK initialized successfully via ADC");
    }

    @Test
    public void test2AuthGetUsers() throws Exception {
        FirebaseAuth auth = FirebaseAuth.getInstance(app);
        assertNotNull(auth);

        try {
            ListUsersPage page = auth.listUsers(null, 1);
            int count = com.google.common.collect.Iterables.size(page.getValues());
            System.out.println("Successfully fetched " + count + " users.");
            assertNotNull(page.getValues());
        } catch (Exception e) {
            System.err.println("Auth test failed: " + e.getMessage());
            throw e;
        }
    }

    @Test
    public void test3FirestoreRead() throws Exception {
        Firestore db = FirestoreClient.getFirestore(app);
        assertNotNull(db);

        try {
            DocumentReference docRef = db.collection("wif-demo").document("test-connection");

            DocumentSnapshot doc = docRef.get().get();
            assertTrue(doc.exists());
            assertEquals("Hello from GitHub Actions WIF!", doc.getString("message"));
            assertNotNull(doc.get("timestamp"));
            System.out.println("Firestore read successful.");
        } catch (Exception e) {
            System.err.println("Firestore test failed: " + e.getMessage());
            throw e;
        }
    }

    @Test
    public void test4CreateCustomToken() throws Exception {
        FirebaseAuth auth = FirebaseAuth.getInstance(app);
        String uid = "wif-demo-user-123";

        try {
            String customToken = auth.createCustomToken(uid);
            assertNotNull(customToken);
            assertFalse(customToken.isEmpty());
            System.out.println("Successfully created a custom signed token.");
        } catch (Exception e) {
            System.err.println("Token signing test failed: " + e.getMessage());
            throw e;
        }
    }
}

File: go/firebase_admin_test.go
package main

import (
	"context"
	"fmt"
	"testing"

	firebase "firebase.google.com/go/v4"
	"google.golang.org/api/iterator"
)

var app *firebase.App

func Test1InitializeApp(t *testing.T) {
	ctx := context.Background()

	config := &firebase.Config{}

	var err error
	app, err = firebase.NewApp(ctx, config)
	if err != nil {
		t.Fatalf("error initializing app: %v", err)
	}

	if app == nil {
		t.Fatal("app is nil")
	}

	fmt.Println("Firebase Admin SDK initialized successfully via ADC")
}

func Test2AuthGetUsers(t *testing.T) {
	ctx := context.Background()
	client, err := app.Auth(ctx)
	if err != nil {
		t.Fatalf("error getting Auth client: %v", err)
	}

	pager := client.Users(ctx, "")
	user, err := pager.Next()
	if err != iterator.Done && err != nil {
		t.Fatalf("error listing users: %v", err)
	}

	if user != nil {
		fmt.Println("Successfully fetched users.")
	} else {
		fmt.Println("No users fetched (iterator empty).")
	}
}

func Test3FirestoreRead(t *testing.T) {
	ctx := context.Background()
	client, err := app.Firestore(ctx)
	if err != nil {
		t.Fatalf("error getting Firestore client: %v", err)
	}
	defer client.Close()

	doc, err := client.Collection("wif-demo").Doc("test-connection").Get(ctx)
	if err != nil {
		t.Fatalf("error reading Firestore document: %v", err)
	}

	if !doc.Exists() {
		t.Fatal("expected document to exist")
	}

	data := doc.Data()
	if data["message"] != "Hello from GitHub Actions WIF!" {
		t.Errorf("unexpected message: %v", data["message"])
	}
	if data["timestamp"] == nil {
		t.Error("expected timestamp to be present")
	}

	fmt.Println("Firestore read successful.")
}

func Test4CreateCustomToken(t *testing.T) {
	ctx := context.Background()
	client, err := app.Auth(ctx)
	if err != nil {
		t.Fatalf("error getting Auth client: %v", err)
	}

	uid := "wif-demo-user-123"
	token, err := client.CustomToken(ctx, uid)
	if err != nil {
		t.Fatalf("error creating custom token: %v", err)
	}

	if token == "" {
		t.Fatal("expected custom token to be non-empty string")
	}

	fmt.Println("Successfully created a custom signed token.")
}
